cr/state/checkpoint.go

```diff
@@ -0,0 +1,183 @@
+package state
+
+import (
+	"io"
+
+	"github.com/elastos/Elastos.ELA/common"
+	"github.com/elastos/Elastos.ELA/utils"
+)
+
+// KeyFrame holds necessary state about state
+type KeyFrame struct {
+	CodeDIDMap         map[string]common.Uint168
+	PendingCandidates  map[common.Uint168]*Candidate
+	ActivityCandidates map[common.Uint168]*Candidate
+	CanceledCandidates map[common.Uint168]*Candidate
+	Nicknames          map[string]struct{}
+}
+
+func (k *KeyFrame) Serialize(w io.Writer) (err error) {
+	if err = k.SerializeCodeAddressMap(w, k.CodeDIDMap); err != nil {
+		return
+	}
+
+	if err = k.SerializeCandidateMap(w, k.PendingCandidates); err != nil {
+		return
+	}
+
+	if err = k.SerializeCandidateMap(w, k.ActivityCandidates); err != nil {
+		return
+	}
+
+	if err = k.SerializeCandidateMap(w, k.CanceledCandidates); err != nil {
+		return
+	}
+
+	if err = utils.SerializeStringSet(w, k.Nicknames); err != nil {
+		return
+	}
+
+	return
+}
+
+func (k *KeyFrame) Deserialize(r io.Reader) (err error) {
+	if k.CodeDIDMap, err = k.DeserializeCodeAddressMap(r); err != nil {
+		return
+	}
+
+	if k.PendingCandidates, err = k.DeserializeCandidateMap(r); err != nil {
+		return
+	}
+
+	if k.ActivityCandidates, err = k.DeserializeCandidateMap(r); err != nil {
+		return
+	}
+
+	if k.CanceledCandidates, err = k.DeserializeCandidateMap(r); err != nil {
+		return
+	}
+
+	if k.Nicknames, err = utils.DeserializeStringSet(r); err != nil {
+		return
+	}
+	return
+}
+
+func (k *KeyFrame) SerializeCodeAddressMap(w io.Writer,
+	cmap map[string]common.Uint168) (err error) {
+	if err = common.WriteVarUint(w, uint64(len(cmap))); err != nil {
+		return
+	}
+	for k, v := range cmap {
+		if err = common.WriteVarString(w, k); err != nil {
+			return
+		}
+
+		if err = v.Serialize(w); err != nil {
+			return
+		}
+	}
+	return
+}
+
+func (k *KeyFrame) DeserializeCodeAddressMap(r io.Reader) (
+	cmap map[string]common.Uint168, err error) {
+	var count uint64
+	if count, err = common.ReadVarUint(r, 0); err != nil {
+		return
+	}
+	cmap = make(map[string]common.Uint168)
+	for i := uint64(0); i < count; i++ {
+		var k string
```

Comment:
```
var k string
```
The variable k is defined outside of the for loop to reduce garbage collection pressure and save memory resources